clients/core/src/main/scala/mqperf/Sender.scala

```diff
@@ -19,49 +21,107 @@ class Sender(config: Config, mq: Mq, clock: Clock) extends StrictLogging {
+      IO.sleep(3.seconds) >>
+      IO.fromFuture(IO(mqSender.close()))
+
+    senderRun
+      .handleError(e => logger.error("Sender failed", e))
+      .unsafeToFuture()
+  }
+
+  private def runParallel(mqSender: MqSender, senderConcurrency: Int): IO[Unit] = {
+    val batchSize = config.batchSizeSend
+
+    def sendRunner(state: Ref[IO, State], iterationNo: Int = 0): IO[Unit] = {
+      def iteration(state: Ref[IO, State]): IO[Unit] = {
+        state.get
+          .flatMap(s =>
+            if (s.sent >= s.permits) {
+              IO.unit
+            } else {
+              for {
+                sendStart <- IO(clock.millis())
+                _ <- IO.fromFuture(IO(mqSender.send(nextMessagesBatch(batchSize))))
+                _ <- state.update { s => s.copy(sent = s.sent + batchSize) }
+                _ <- IO {
+                  LocalMetrics.messageCounter.inc(batchSize.toDouble)
+                  LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
+                }
+                result <- iteration(state)
+              } yield result
             }
+          )
+          .handleErrorWith { e =>
+            IO(logger.error(s"Sender iteration[$iterationNo] failed ", e)) >>
+              iteration(state)
           }
+      }
 
-          val iterationEnd = clock.millis()
-          // a whole iteration should take at least a second
-          Thread.sleep(math.max(0, 1000L - (iterationEnd - iterationStart)))
+      state.get.flatMap(s => {
+        if (iterationNo < s.currentIteration) {
+          iteration(state) >> sendRunner(state, iterationNo + 1)
+        } else {
+          s.nextIterationHook.get.flatMap(isTestEnd => {
+            if (isTestEnd) IO.unit
+            else iteration(state) >> sendRunner(state, iterationNo + 1)
+          })
         }
+      })
+    }
 
-        logger.info("Sending done, waiting for all messages to be flushed ...")
-        Thread.sleep(3000L)
+    def loop(until: Long, state: Ref[IO, State], iterationNo: Int = 0): IO[Unit] = {
+      if (clock.millis() >= until) {
+        state.get
+          .flatMap(_.nextIterationHook.complete(true))
+          .void
+      } else {
+        for {
+          nextIterationH <- Deferred[IO, Boolean]
+          _ <- state.get.flatMap(_.nextIterationHook.complete(false))
+          addedPermits <- state.modify(s => {
+            val p = newPermits(s.sent, s.permits)
+            s.copy(
+              permits = s.permits + p,
+              currentIteration = s.currentIteration + 1,
+              nextIterationHook = nextIterationH
+            ) -> p
+          })
+          _ <- IO(logger.info(s"Messages to send: $addedPermits (concurrency=${config.senderConcurrency})"))
+          _ <- IO.sleep(1.second)
+          result <- loop(until, state, iterationNo + 1)
+        } yield result
       }
-    }.flatMap(_ => mqSender.close())
+    }
+
+    def newPermits(sent: Long, permits: Long) = {
```

Comment:
I'm not sure if this will send at most `config.msgsPerSecond`. If in second 1 we send 1 message, then at second 2 we'll add the whole `config.msgsPerSecond`, so in second 2 we'll be able to send `2*msgsPerSecond-1`